atlasdb-commons/src/main/java/com/palantir/common/compression/BufferedDelegateInputStream.java

```diff
@@ -0,0 +1,127 @@
+
+    public BufferedDelegateInputStream(InputStream delegate, int bufferLength) {
+        Preconditions.checkArgument(bufferLength >= 0, "buffer size must be greater than or equal to zero");
+        this.delegate = delegate;
+        this.position = 0;
+        this.bufferSize = 0;
+        this.buffer = new byte[bufferLength];
+    }
+
+    @Override
+    public final int read() throws IOException {
+        if (!ensureBytesAvailable()) {
+            return READ_FAILED;
+        }
+        return buffer[position++] & 0xFF;
+    }
+
+    // If all bytes in the buffer have been read, attempt to refill
+    // the buffer. If the buffer can't be refilled, then we've reached
+    // the end of the stream and return false.
+    private boolean ensureBytesAvailable() throws IOException {
+        if (position == bufferSize) {
+            boolean bufferRefilled = refill();
+            if (!bufferRefilled) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Reads up to {@code len} bytes into the provided byte array {@code b}.
+     * If no remaining unread bytes are available in the internal buffer,
+     * {@link #refill() refill} is called to refill the internal buffer.
+     * Otherwise, either {@code len} bytes or the number of bytes available
+     * in the buffer are read, whichever is smaller.
+     *
+     * To read from the buffer until either the end of the stream is reached
+     * or {@code len} bytes have been read, use
+     * {@link com.google.common.io.ByteStreams#read(InputStream, byte[], int, int)
+     * ByteStreams.read}.
+     */
+    @Override
+    public final int read(byte[] b, int off, int len) throws IOException {
+        if (b == null) {
+            throw new NullPointerException("Provided byte array b cannot be null.");
+        } else if (off < 0 || len < 0 || len > b.length - off) {
+            throw new IndexOutOfBoundsException();
+        } else if (len == 0) {
+            return 0;
+        }
+
+        if (!ensureBytesAvailable()) {
+            return READ_FAILED;
+        }
+
+        return readBytes(b, off, len);
+    }
+
+    private int readBytes(byte[] dest, int offset, int maxBytesToRead) {
+        int bytesLeftInBuffer = available();
+        int bytesToRead = Math.min(maxBytesToRead, bytesLeftInBuffer);
+        System.arraycopy(buffer, position, dest, offset, bytesToRead);
+        position += bytesToRead;
+        return bytesToRead;
+    }
+
+    @Override
+    public final int available() {
+        return bufferSize - position;
+    }
+
+    @Override
+    public void close() throws IOException {
+        delegate.close();
+    }
+
+    /**
+     * Refills the internal buffer from the delegate {@link InputStream}.
+     *
+     * @return True if the buffer was refilled. False if the delegate
+     *         stream has been exhausted and the internal buffer is empty.
+     */
+    protected abstract boolean refill() throws IOException;
```

Comment:
although this would make `LZ4CompressingInputStream` more complicated (you'd probably have to do another copy or set the buffer on the `InternalOutputStream` in the `refill` call)..